### PR TITLE
Code coverage support

### PR DESCRIPTION
**Pull request by DinoV**
Adds support - both for running tests with code coverage as well as importing a .coveragexml file from coverage.py.

### CODE REVIEW

File: Python/Product/TestAdapter/visualstudio_py_testlauncher.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ def main():
     parser.add_option('-x', '--mixed-mode', action='store_true', help='wait for mixed-mode debugger to attach')
     parser.add_option('-t', '--test', type='str', dest='tests', action='append', help='specifies a test to run')
     parser.add_option('-m', '--module', type='str', help='name of the module to import the tests from')
-    
+    parser.add_option('-c', '--coverage', type='str', help='enable code coverage and speicfy filename')
```
Comment (huguesv) [inline Python/Product/TestAdapter/visualstudio_py_testlauncher.py]: typo

File: Python/Product/PythonTools/PythonTools/CodeCoverage/CoverageFileInfo.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+
+namespace Microsoft.PythonTools.CodeCoverage {
+    class CoverageFileInfo {
+        public readonly string Filename;
+        public readonly HashSet<int> Hits;
+
+        public CoverageFileInfo(string Filename, HashSet<int> hits) {
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/CodeCoverage/CoverageFileInfo.cs]: lowercase

File: Python/Product/PythonTools/PythonTools/CodeCoverage/CoveragePyConverter.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        private string GetFilename(string basePath, XPathNavigator node) {
+            // Try and find the source relative to the coverage file first...
+            var filename = node.GetAttribute("filename", "");
+            string relativePath = Path.Combine(_coverageXmlBasePath, filename);
+            if (File.Exists(relativePath)) {
+                return relativePath;
+            }
+
+            // Then try the absolute path.
+            return Path.Combine(basePath, node.GetAttribute("filename", ""));
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/CodeCoverage/CoveragePyConverter.cs]: already in filename variable

File: Python/Product/PythonTools/PythonTools/CodeCoverage/CoverageScope.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+using Microsoft.PythonTools.Parsing;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.CodeCoverage {
+    /// <summary>
+    /// Tracks data about coverage for each scope.
+    /// </summary>
+
+    sealed class CoverageScope {
+        public ScopeStatement Statement;
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/CodeCoverage/CoverageScope.cs]: Can most of these be readonly?

Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/CodeCoverage/CoverageScope.cs]: Seems like it would be better to use properties too.

---
In reply to: [66358430](https://github.com/Microsoft/PTVS/pull/1337#discussion_r66358430) [](ancestors = 66358430)

Comment (DinoV) [inline Python/Product/PythonTools/PythonTools/CodeCoverage/CoverageScope.cs]: It's just an internal class so there's no need for the overhead of properties

File: Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.PythonTools.CodeCoverage;
+using Microsoft.PythonTools.Parsing;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudioTools;
+
+namespace Microsoft.PythonTools.Commands {
+    /// <summary>
+    /// Provides the command for starting a file or the start item of a project in the REPL window.
```
Comment (huguesv) [inline Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs]: wrong description ;)

**Comment by DinoV** [comment]
Fixes #1020

**Comment by DinoV** [comment]
Fixes #1113

**Comment by DinoV** [comment]
Now fixes #1067 and #213

File: Python/Product/Analyzer/Intellisense/AnalysisProtocol.cs [hunk 6]
```diff
@@ -851,5 +851,50 @@ internal class UnhandledExceptionEvent : Event {
             }
             public override string name => Name;
         }
+
+        public sealed class GetTestCasesRequest : Request<GetTestCasesResponse> {
+            public const string Command = "getTestCases";
+
+            public override string command => Command;
+
+            public string filename;
```
Comment (zooba) [inline Python/Product/Analyzer/Intellisense/AnalysisProtocol.cs]: Guessing we can extend this easily/safely in the future? If, say, we wanted to add alternate detection methods other than `unittest`?

File: Python/Product/Ipc.Json/Connection.cs [hunk 7]
```diff
@@ -131,7 +131,7 @@ public sealed class Connection : IDisposable {
         /// <summary>
         /// Returns a task which will process incoming messages.  This can be started on another thread or
         /// in whatever form of synchronization context you like.  StartProcessing is a convenience helper
-        /// for startin this running asynchronously using Test.Run.
+        /// for startin this running asynchronously using Task.Run.
```
Comment (zooba) [inline Python/Product/Ipc.Json/Connection.cs]: starting*

Comment (zooba) [inline Python/Product/Ipc.Json/Connection.cs]: Alternatively, `startin'` ;)

File: Python/Product/PythonTools/PythonTools.csproj [hunk 8]
```diff
@@ -170,6 +170,17 @@
...
+      <HintPath>$(DevEnvDir)\CommonExtensions\Microsoft\TestWindow\Microsoft.VisualStudio.TestPlatform.ObjectModel.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="Microsoft.VisualStudio.TestWindow.CodeCoverage">
+      <HintPath>$(DevEnvDir)\CommonExtensions\Microsoft\AnalyzeCodeCoverage\Microsoft.VisualStudio.TestWindow.CodeCoverage.dll</HintPath>
+    </Reference>
+    <Reference Include="Microsoft.VisualStudio.TestWindow.Interfaces">
+      <HintPath>$(DevEnvDir)\CommonExtensions\Microsoft\TestWindow\Microsoft.VisualStudio.TestWindow.Interfaces.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools.csproj]: Any chance we can do this without depending on these assemblies from PythonTools? They may not be available on all VS installs, and I'd much rather degrade gracefully than crash on startup.

File: Python/Product/PythonTools/PythonTools/CodeCoverage/CoveragePyConverter.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        public CoverageFileInfo[] Parse() {
+            XmlDocument doc = new XmlDocument();
+            doc.Load(_input);
+            Dictionary<string, HashSet<int>> data = new Dictionary<string, HashSet<int>>();
+
+            var root = doc.DocumentElement.CreateNavigator();
+            string basePath = "";
+            foreach (XPathNavigator source in root.Select("/coverage/sources/source")) {
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/CodeCoverage/CoveragePyConverter.cs]: What happens if this path is not found? Do we crash or just not return any results?

File: Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            if (file == null) {
+                file = _serviceProvider.BrowseForFileOpen(
+                    IntPtr.Zero,
+                    "Coverage XML (*.xml)|*.xml|All Files (*.*)|*.*"
+                );
+            }
+
+            if (file != null) {
+                var outFilename = Path.Combine(
+                    Path.GetDirectoryName(file),
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs]: There are `PathUtils` equivalents for these that are safer (handle long paths, etc.)

File: Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+                    IntPtr.Zero,
+                    "Coverage XML (*.xml)|*.xml|All Files (*.*)|*.*"
+                );
+            }
+
+            if (file != null) {
+                var outFilename = Path.Combine(
+                    Path.GetDirectoryName(file), 
+                    Path.GetFileNameWithoutExtension(file) + ".coveragexml"
+                );
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs]: Actually, looks like the whole statement is equivalent to `Path.ChangeExtension(file, ".coveragexml")`? Which also handles long paths fine.

File: Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+                ConvertCoveragePy(file, outFilename, version);
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs]: Would be nice to gracefully handle `IOException`s here. Not sure what the best behavior would be necessarily, but almost anything is better than "go read ActivityLog.xml", which is what people will get if (e.g.) the output file already exists and is locked/hidden.

File: Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+                TestContainer container;
+                if (analysis != null && projectInfo._containers.TryGetValue(analysis, out container)) {
+                    return container;
+                }
+            }
+
+            return null;
+        }
+
+        private static IEnumerable<IVsProject> EnumerateLoadedProjects(IVsSolution solution) {
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs]: There's already an extension method on `IVsSolution` for this (in `PythonTools` project), but if we move this class back to its own assembly then I guess this has to stay.

File: Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+                _project = project;
+                _discoverer = discoverer;
+                _containers = new Dictionary<AnalysisEntry, TestContainer>();
+
+                project.ProjectAnalyzerChanged += ProjectAnalyzerChanged;
+                HookAnalysisComplete();
+            }
+
+            private void HookAnalysisComplete() {
+                _project.GetAnalyzer().AnalysisComplete += AnalysisComplete;
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs]: Where do we detach this event?

File: Python/Product/TestAdapter/TestAdapter.csproj [hunk 15]
```diff
@@ -114,28 +117,18 @@
...
-      <Link>TestFileAddRemoveListener.cs</Link>
-    </Compile>
-    <Compile Include="..\..\..\Common\Product\TestAdapter\TestFilesUpdateWatcher.cs">
-      <Link>TestFilesUpdateWatcher.cs</Link>
-    </Compile>
-    <Compile Include="TestExecutorProjectContext.cs" />
-    <Compile Include="TestMethodResolver.cs" />
-    <Compile Include="..\..\..\Common\Product\TestAdapter\VsProjectExtensions.cs">
-      <Link>VsProjectExtensions.cs</Link>
-    </Compile>
```
Comment (zooba) [inline Python/Product/TestAdapter/TestAdapter.csproj]: Feels like we've divided a lot of test functionality between two assemblies now... I'd much rather keep it loosely coupled in TestAdapter.dll so that it's easy to drop completely when unavailable.

Comment (zooba) [inline Python/Product/TestAdapter/TestAdapter.csproj]: Or to substitute for a different discoverer/executor implementation, while still reusing whatever core Python functionality (analysis/coverage.py/etc.) can be reliably added into the core DLL. Or better yet, its own assemblies, so that we can keep startup time down by lazily loading components as much as possible.

File: Python/Product/TestAdapter/TestExecutor.cs [hunk 16]
```diff
@@ -85,6 +81,52 @@ class TestExecutor : ITestExecutor {
...
+                if (!String.IsNullOrWhiteSpace(djangoSettings)) {
+                    projSettings.Environment["DJANGO_SETTINGS_MODULE"] = djangoSettings;
+                }
+
+                foreach (XPathNavigator searchPath in project.Select("SearchPaths/Search")) {
+                    projSettings.SearchPath.Add(searchPath.GetAttribute("value", ""));
+                }
+
+                foreach (XPathNavigator test in project.Select("Test")) {
+                    res[test.GetAttribute("file", "")] = projSettings;
```
Comment (zooba) [inline Python/Product/TestAdapter/TestExecutor.cs]: Do we really want to have `res[""]` set to the last Test with no filename included? Or should we just skip tests that don't include a filename? (Should be possible to encode that into XPath with something like `"Test[string-length(@file) > 0]"`)

File: Python/Product/TestAdapter/TestExecutor.cs [hunk 17]
```diff
@@ -100,293 +142,568 @@ class TestExecutor : ITestExecutor {
...
                 }
 
-                try {
-                    RunTestCase(frameworkHandle, runContext, test, sourceToSettings);
-                } catch (Exception ex) {
-                    frameworkHandle.SendMessage(TestMessageLevel.Error, ex.ToString());
+                var runner = new TestRunner(
+                    frameworkHandle,
+                    runContext,
+                    tests,
```
Comment (zooba) [inline Python/Product/TestAdapter/TestExecutor.cs]: Shouldn't we pass `testGroup` here? Rather than all tests

File: Python/Product/TestAdapter/visualstudio_py_testlauncher.py [hunk 18]
```diff
@@ -17,23 +17,174 @@
...
+        if self.old_out:
+            self.old_out.flush()
+    
+    def writelines(self, lines):
+        for line in lines:
+            self.write(line)
+    
+    @property
+    def encoding(self):
+        return 'utf8'
```
Comment (zooba) [inline Python/Product/TestAdapter/visualstudio_py_testlauncher.py]: Just assign `_TestOutput.encoding`?

File: Python/Product/TestAdapter/visualstudio_py_testlauncher.py [hunk 19]
```diff
@@ -17,23 +17,174 @@
...
+
+    def next(self):
+        pass
+    
+    @property
+    def name(self):
+        if self.is_stdout:
+            return "<stdout>"
+        else:
+            return "<stderr>"
```
Comment (zooba) [inline Python/Product/TestAdapter/visualstudio_py_testlauncher.py]: Just assign `self.name`?

File: Python/Product/TestAdapter/visualstudio_py_testlauncher.py [hunk 20]
```diff
@@ -17,23 +17,174 @@
...
+    parser.add_option('-c', '--coverage', type='str', help='enable code coverage and specify filename')
+    parser.add_option('-r', '--result-port', type='int', help='connect to port on localhost and send test results')
     (opts, _) = parser.parse_args()
     
     sys.path[0] = os.getcwd()
     
+    if opts.result_port:
+        _channel = _IpcChannel(socket.create_connection(('127.0.0.1', opts.result_port)))
+        sys.stdout = _TestOutput(sys.stdout, is_stdout = True)
+        sys.stderr = _TestOutput(sys.stderr, is_stdout = False)
```
Comment (zooba) [inline Python/Product/TestAdapter/visualstudio_py_testlauncher.py]: No spaces around `=` for named arguments

Comment (zooba) [inline Python/Product/TestAdapter/visualstudio_py_testlauncher.py]: Might be better to pass in a `name` argument anyway. Certainly a nicer looking API

File: Python/Tests/TestAdapterTests/TestDiscovererTests.cs [hunk 21]
```diff
@@ -35,35 +38,45 @@ public class TestDiscovererTests {
...
+<Test className=""Test_test1"" file=""C:\Visual Studio 2015\Projects\PythonApplication107\PythonApplication107\test1.py"" line=""17"" column=""9"" method=""test_A"" />
+<Test className=""Test_test1"" file=""C:\Visual Studio 2015\Projects\PythonApplication107\PythonApplication107\test1.py"" line=""21"" column=""9"" method=""test_B"" />
+<Test className=""Test_test2"" file=""C:\Visual Studio 2015\Projects\PythonApplication107\PythonApplication107\test1.py"" line=""48"" column=""9"" method=""test_C"" /></Project></TestCases></Python></RunSettings>";
```
Comment (zooba) [inline Python/Tests/TestAdapterTests/TestDiscovererTests.cs]: Interesting choice of filename... you're not relying on something on your own system are you?

If the file needs to actually exist, use `TestData.GetTempPath` and write it to there.

File: Python/Tests/TestAdapterTests/TestExecutorTests.cs [hunk 22]
```diff
@@ -59,6 +59,21 @@ public class TestExecutorTests {
...
+            var file1 = "C:\\Some\\Path\\file1.py";
+            var file2 = "C:\\Some\\Path\\file2.py";
+            var best = TestExecutor.UpdateBestFile(null, file1);
+            Assert.AreEqual(best, file1);
+
+            best = TestExecutor.UpdateBestFile(null, file1);
+            Assert.AreEqual(best, file1);
+
+            best = TestExecutor.UpdateBestFile(best, file2);
+            Assert.AreEqual("C:\\Some\\Path", best);
```
Comment (zooba) [inline Python/Tests/TestAdapterTests/TestExecutorTests.cs]: I feel like there are probably more edge cases to test here...

File: Python/Tests/TestData/Coverage/MultiModule/coverage.xml [hunk 23]
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" ?>
+<coverage branch-rate="0" line-rate="0.9697" timestamp="1465324878063" version="4.1">
+	<!-- Generated by coverage.py: https://coverage.readthedocs.io -->
+	<!-- Based on https://raw.githubusercontent.com/cobertura/web/f0366e5e2cf18f111cbd61fc34ef720a6584ba02/htdocs/xml/coverage-03.dtd -->
+	<sources>
+		<source>C:\Users\dinov\Documents\Visual Studio 2015\Projects\PythonApplication50\PythonApplication50\env2</source>
```
Comment (zooba) [inline Python/Tests/TestData/Coverage/MultiModule/coverage.xml]: Not sure if this will affect any of our handling/parsing?

File: Python/Tests/TestData/Coverage/MultiModule/foo/__init__.py [hunk 24]
```diff
@@ -0,0 +1,2 @@
+import blah
+print('hi')
```
Comment (zooba) [inline Python/Tests/TestData/Coverage/MultiModule/foo/__init__.py]: You've started writing Python 3 code :)

File: Python/Product/Analyzer/Intellisense/AnalysisProtocol.cs (hunk 6 above)
Comment (DinoV) [inline Python/Product/Analyzer/Intellisense/AnalysisProtocol.cs]: Yep, this is all an internal implementation detail


---
In reply to: [67382713](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67382713) [](ancestors = 67382713)

File: Python/Tests/TestData/Coverage/Statements/test.py [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+    x = 1
+except:
+    pass
+    
+try:
+    x = 1
+finally:
+    pass
+    
+with open('foo.txt', 'w+'):
```
Comment (zooba) [inline Python/Tests/TestData/Coverage/Statements/test.py]: We can't use `foo` in our code without upsetting the censors. `fob` or `spam` are better choices.

File: Python/Product/PythonTools/PythonTools/CodeCoverage/CoveragePyConverter.cs (hunk 9 above)
Comment (DinoV) [inline Python/Product/PythonTools/PythonTools/CodeCoverage/CoveragePyConverter.cs]: It should return an empty enumerable 

---
In reply to: [67383482](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67383482) [](ancestors = 67383482)

File: Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs (hunk 10 above)
Comment (DinoV) [inline Python/Product/PythonTools/PythonTools/Commands/ImportCoverageCommand.cs]: Which calls are you suggesting should be replaced w/ PathUtils calls?  PathUtils doesn't have Combine, GetDirectoryName, or GetFileNameWithoutExtension.  And long paths don't really matter here because we're not going to be able to write to them anyway.

---
In reply to: [67383829](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67383829) [](ancestors = 67383829)

File: Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs (hunk 14 above)
Comment (DinoV) [inline Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs]: In OnProjectUnloaded

---
In reply to: [67384909](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67384909) [](ancestors = 67384909)

File: Python/Product/TestAdapter/TestExecutor.cs (hunk 16 above)
Comment (DinoV) [inline Python/Product/TestAdapter/TestExecutor.cs]: We're on the other side of this generating this XML - so we'll actually never have an empty file.  I'll add an assertion

---
In reply to: [67386339](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67386339) [](ancestors = 67386339)

File: Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs (hunk 14 above)
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs]: What if the analyzer changes?

---
In reply to: [67420732](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67420732) [](ancestors = 67420732,67384909)

File: Python/Product/TestAdapter/TestExecutor.cs (hunk 17 above)
Comment (DinoV) [inline Python/Product/TestAdapter/TestExecutor.cs]: Yes!


---
In reply to: [67386465](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67386465) [](ancestors = 67386465)

File: Python/Product/TestAdapter/TestExecutor.cs (hunk 16 above)
Comment (zooba) [inline Python/Product/TestAdapter/TestExecutor.cs]: And some fallback behaviour - crash if we need to, but with something that makes it obvious that an assertion failed in the release build..

---
In reply to: [67421586](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67421586) [](ancestors = 67421586,67386339)

File: Python/Tests/TestAdapterTests/TestDiscovererTests.cs (hunk 21 above)
Comment (DinoV) [inline Python/Tests/TestAdapterTests/TestDiscovererTests.cs]: The file doesn't actually need to exist, this is just testing that we're able to parse out the XML appropriately, these filenames could be whatever...  I just cleaned them up from where they were originally generated on my system.

---
In reply to: [67395057](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67395057) [](ancestors = 67395057)

File: Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs (hunk 14 above)
Comment (DinoV) [inline Python/Product/PythonTools/PythonTools/TestAdapter/TestContainerDiscoverer.cs]: If the analyzer changes it's going to be collected and go away :)

---
In reply to: [67421610](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67421610) [](ancestors = 67421610,67420732,67384909)

File: Python/Product/Analysis/Projects/PythonProject.cs [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+        public abstract string GetUnevaluatedProperty(string name);
+
+        /// <summary>
+        /// Sets a property for the project.  See GetProperty for more information on common properties.
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="value"></param>
+        public abstract void SetProperty(string name, string value);
+
+        public abstract IPythonInterpreterFactory GetInterpreterFactory();
```
Comment (zooba) [inline Python/Product/Analysis/Projects/PythonProject.cs]: Is the default behaviour of this going to be to throw if it can't be returned? If so, rename GetLaunchConfigurationOrThrow to remove the "or throw", since it's just part of the regular behaviour.

**Comment by zooba** [comment]
> using System;

Header comment

---
Refers to: Python/Product/Analyzer/Intellisense/FileExtensionAttribute.cs:1 in d396cfb. [](commit_id = d396cfb35e36ac95c698f7643b6085a6499cd26a, deletion_comment = False)

File: Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs [hunk 27]
```diff
@@ -35,6 +35,7 @@
 using Microsoft.PythonTools.Parsing.Ast;
 
 namespace Microsoft.PythonTools.Intellisense {
+    using Projects;
```
Comment (zooba) [inline Python/Product/Analyzer/Intellisense/OutOfProcProjectAnalyzer.cs]: Can we put this up with the other usings?

File: Python/Product/PythonTools/PythonTools/Intellisense/ProjectAnalyzer.cs [hunk 28]
```diff
@@ -217,6 +218,20 @@ out _analysisProcess
...
         }
 
+        /// <summary>
+        /// Raised when any file has a new analysis.
+        /// </summary>
+        public override event EventHandler<AnalysisCompleteEventArgs> AnalysisComplete;
+
+        public override IEnumerable<string> Files {
+            get {
+                return _projectFiles.Keys;
```
Comment (zooba) [inline Python/Product/PythonTools/PythonTools/Intellisense/ProjectAnalyzer.cs]: Do we need to lock and enumerate this to avoid race conditions?

File: Python/Product/TestAdapter/PythonRunSettings.cs [hunk 29]
```diff
@@ -0,0 +1,227 @@
...
+
+        public IXPathNavigable AddRunSettings(IXPathNavigable inputRunSettingDocument, IRunSettingsConfigurationInfo configurationInfo, ILogger log) {
+            XPathNavigator navigator = inputRunSettingDocument.CreateNavigator();
+            var python = navigator.Select("/RunSettings");
+            if (python.MoveNext()) {
+                using (var writer = python.Current.AppendChild()) {
+                    var pyContainers = configurationInfo.TestContainers
+                        .Where(x => x is TestContainer)
+                        .Cast<TestContainer>()
+                        .GroupBy(x => x.Project);
```
Comment (zooba) [inline Python/Product/TestAdapter/PythonRunSettings.cs]: .OfType<TestContainer>()

Comment (zooba) [inline Python/Product/TestAdapter/PythonRunSettings.cs]: Instead of .Where().Cast<T>()

---
In reply to: [68104358](https://github.com/Microsoft/PTVS/pull/1337#discussion_r68104358) [](ancestors = 68104358)

File: Python/Product/TestAdapter/PythonRunSettings.cs [hunk 30]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+        private void StateChange(object sender, OperationStateChangedEventArgs e) {
+            if (e.State == TestOperationStates.TestExecutionFinished) {
+                var resultUris = e.Operation.GetRunSettingsDataCollectorResultUri(PythonCodeCoverageUri);
+                if (resultUris != null) {
+                    foreach (var eachAttachment in resultUris) {
+                        string filePath = eachAttachment.LocalPath;
+
+                        if (!string.IsNullOrEmpty(filePath) && File.Exists(filePath)) {
```
Comment (zooba) [inline Python/Product/TestAdapter/PythonRunSettings.cs]: File.Exists handles null and empty strings just fine - no need for an extra test

File: Python/Product/TestAdapter/PythonRunSettings.cs [hunk 31]
```diff
@@ -0,0 +1,227 @@
...
+                        foreach (var container in project) {
+                            writer.WriteStartElement("Project");
+                            writer.WriteAttributeString("home", project.Key.ProjectHome);
+
+                            LaunchConfiguration config = null;
+                            string nativeCode = "", djangoSettings = "";
+                            _dispatcher.Invoke(() => {
+                                try {
+                                    config = project.Key.GetLaunchConfigurationOrThrow();
+                                } catch {
```
Comment (zooba) [inline Python/Product/TestAdapter/PythonRunSettings.cs]: If this happens, we're just going to crash below. We need to do something much more sensible when we can't get a launch configuration from the project.

File: Python/Product/TestAdapter/TestContainerDiscoverer.cs [hunk 32]
```diff
@@ -167,330 +135,121 @@ private TestContainerDiscoverer([Import(typeof(SVsServiceProvider))]IServiceProv
...
-                    // VS renames files instead of overwriting them when
-                    // saving, so we need to listen for renames where the new
-                    // path is part of the project.
-                    case TestFileChangedReason.Renamed:
-                    case TestFileChangedReason.Changed:
-                        OnTestContainersChanged(GetTestProjectFromFile(e.File));
-                        break;
+            private Architecture Architecture {
+                get {
+                    return Architecture.X86;
```
Comment (zooba) [inline Python/Product/TestAdapter/TestContainerDiscoverer.cs]: I suspect we can switch to Any here and then we'll work with 64-bit tests, but we can try that later.

**Comment by zooba** [comment]
> namespace Microsoft.PythonTools.Analysis {

Header comment

---
Refers to: Python/Product/TestAdapter.Analysis/TestCaseInfo.cs:1 in d396cfb. [](commit_id = d396cfb35e36ac95c698f7643b6085a6499cd26a, deletion_comment = False)

**Comment by zooba** [comment]
> namespace Microsoft.PythonTools.Analysis {

And fix the namespace

---
In reply to: [227834719](https://github.com/Microsoft/PTVS/pull/1337#issuecomment-227834719) [](ancestors = 227834719)

---
Refers to: Python/Product/TestAdapter.Analysis/TestCaseInfo.cs:1 in d396cfb. [](commit_id = d396cfb35e36ac95c698f7643b6085a6499cd26a, deletion_comment = False)

File: Python/Product/Analysis/Projects/PythonProject.cs (hunk 26 above)
Comment (DinoV) [inline Python/Product/Analysis/Projects/PythonProject.cs]: Nope, it always returns an interpreter factory, falling back to the default one.

---
In reply to: [68103466](https://github.com/Microsoft/PTVS/pull/1337#discussion_r68103466) [](ancestors = 68103466)

File: Python/Product/PythonTools/PythonTools/Intellisense/ProjectAnalyzer.cs (hunk 28 above)
Comment (DinoV) [inline Python/Product/PythonTools/PythonTools/Intellisense/ProjectAnalyzer.cs]: It's a concurrent dictionary

---
In reply to: [68104072](https://github.com/Microsoft/PTVS/pull/1337#discussion_r68104072) [](ancestors = 68104072)
